File: experiments/cursor_sdk/src/knowledgeBase.ts
import fs from "node:fs";
import path from "node:path";

import { parse } from "csv-parse/sync";

import { KNOWLEDGE_BASE_DIR } from "./config.js";
import { parsePipeList, readJsonFile } from "./utils.js";

export interface KnowledgeBaseEntry {
  specialty_name: string;
  folder_name: string;
  disease_catalog: string;
  drug_catalog: string;
  risk_rules: string;
}

const SPECIALTY_ALIASES: Record<string, string> = {
  心内: "心血管",
  心血管: "心血管",
  神经: "神经",
  呼吸: "呼吸",
  肾内: "肾内/泌尿",
  泌尿: "肾内/泌尿",
  肾内泌尿: "肾内/泌尿",
  "肾内/泌尿": "肾内/泌尿",
  内分泌: "内分泌/代谢",
  代谢: "内分泌/代谢",
  内分泌代谢: "内分泌/代谢",
  "内分泌/代谢": "内分泌/代谢",
  消化: "消化",
};

function readCsv(filePath: string): Array<Record<string, string>> {
  const content = fs.readFileSync(filePath, "utf-8");
  return parse(content, {
    columns: true,
    skip_empty_lines: true,
  }) as Array<Record<string, string>>;
}

function resolveKnowledgePath(rawPath: string, folderName: string): string {
  if (fs.existsSync(rawPath)) {
    return rawPath;
  }
  if (!path.isAbsolute(rawPath)) {
    const repoPath = path.join(KNOWLEDGE_BASE_DIR, "..", rawPath);
    if (fs.existsSync(repoPath)) {
      return repoPath;
    }
  }

  const fileName = path.win32.basename(rawPath);
  return path.join(KNOWLEDGE_BASE_DIR, folderName, fileName);
}

function normalizeSpecialtyName(value: string): string {
  const normalized = value.replace(/\uFEFF/g, "").trim();
  const compact = normalized.replace(/\s+/g, "");
  return SPECIALTY_ALIASES[compact] ?? normalized;
}

export class KnowledgeBaseIndex {
  private readonly rows: KnowledgeBaseEntry[];

  constructor() {
    this.rows = readCsv(path.join(KNOWLEDGE_BASE_DIR, "kb_index.csv")).map((row) => ({
      specialty_name: normalizeSpecialtyName(row.specialty_name ?? row["\uFEFFspecialty_name"] ?? ""),
      folder_name: row.folder_name,
      disease_catalog: resolveKnowledgePath(row.disease_catalog, row.folder_name),
      drug_catalog: resolveKnowledgePath(row.drug_catalog, row.folder_name),
      risk_rules: resolveKnowledgePath(row.risk_rules, row.folder_name),
    }));
  }

  getEntry(specialtyName: string): KnowledgeBaseEntry {
    const normalizedName = normalizeSpecialtyName(specialtyName);
    const entry = this.rows.find((row) => row.specialty_name === normalizedName);
    if (!entry) {
      throw new Error(`未找到专科知识库索引: ${specialtyName}`);
    }
    return entry;
  }
}

export function loadSpecialtyKnowledge(specialtyName: string) {
  const kbIndex = new KnowledgeBaseIndex();
  const entry = kbIndex.getEntry(specialtyName);
  const diseaseCatalog = readCsv(entry.disease_catalog);
  const drugCatalog = readCsv(entry.drug_catalog);
  const riskRules = readJsonFile<Array<Record<string, unknown>>>(entry.risk_rules);

  const diagnosticKnowledge = diseaseCatalog.slice(0, 30).map((row) => ({
    disease_name: row.disease_name,
    aliases: parsePipeList(row.aliases),
    diagnostic_basis: row.diagnostic_basis,
    key_symptoms: parsePipeList(row.key_symptoms),
    key_labs_or_tests: parsePipeList(row.key_labs_or_tests),
    differential_diagnosis: parsePipeList(row.differential_diagnosis),
    reference_source: row.reference_source,
    reference_url: row.reference_url,
    agent_use: row.agent_use,
  }));
  const diseaseDirectedDrugs = drugCatalog
    .filter((row) => ["disease_directed_therapy", "risk_modifying_therapy"].includes(row.treatment_role))
    .slice(0, 30)
    .map((row) => ({
      standard_drug_name: row.standard_drug_name,
      aliases: parsePipeList(row.aliases),
      drug_class: row.drug_class,
      disease_context: parsePipeList(row.disease_context),
      treatment_role: row.treatment_role,
      order_category: row.order_category,
      mechanism_or_function: row.mechanism_or_function,
      major_cautions: row.major_cautions,
      reference_source: row.reference_source,
      reference_url: row.reference_url,
    }));
  const supportiveDrugs = drugCatalog
    .filter((row) =>
      ["supportive_or_symptomatic_therapy", "general_inpatient_medication"].includes(row.treatment_role),
    )
    .slice(0, 20)
    .map((row) => ({
      standard_drug_name: row.standard_drug_name,
      aliases: parsePipeList(row.aliases),
      drug_class: row.drug_class,
      treatment_role: row.treatment_role,
      order_category: row.order_category,
      mechanism_or_function: row.mechanism_or_function,
      major_cautions: row.major_cautions,
      reference_source: row.reference_source,
      reference_url: row.reference_url,
    }));
  const summarizedRiskRules = riskRules.slice(0, 8).map((rule) => ({
    rule_id: rule.rule_id,
    risk_target: rule.risk_target,
    related_treatments: rule.related_treatments,
    lab_name: rule.lab_name,
    threshold_type: rule.threshold_type,
    moderate_risk_threshold: rule.moderate_risk_threshold,
    high_risk_threshold: rule.high_risk_threshold,
    risk_message: rule.risk_message,
    action: rule.action,
    contraindication_or_caution: rule.contraindication_or_caution,
    monitoring_advice: rule.monitoring_advice,
    reference_source: rule.reference_source,
    reference_url: rule.reference_url,
  }));

  return {
    entry,
    diseaseCatalog,
    drugCatalog,
    riskRules,
    promptPayload: {
      specialty_name: specialtyName,
      knowledge_base_dir: entry.folder_name,
      diagnostic_knowledge: diagnosticKnowledge,
      drug_function_knowledge: {
        disease_directed_or_risk_modifying: diseaseDirectedDrugs,
        supportive_or_general: supportiveDrugs,
      },
      risk_rules: summarizedRiskRules,
    },
  };
}

export function extractAvoidDrugs(rows: Array<Record<string, string>>): string[] {
  const values = rows
    .filter((row) =>
      ["supportive_or_symptomatic_therapy", "general_inpatient_medication"].includes(row.treatment_role),
    )
    .flatMap((row) => [row.standard_drug_name, ...parsePipeList(row.aliases)])
    .filter(Boolean);
  return Array.from(new Set(values));
}

File: experiments/knowledge_base.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path, PureWindowsPath
from typing import Any

import pandas as pd

from experiments.config import ExperimentPaths


def read_csv_flexible(path: Path) -> pd.DataFrame:
    encodings = ["utf-8-sig", "utf-8", "gb18030"]
    last_error = None
    for encoding in encodings:
        try:
            return pd.read_csv(path, encoding=encoding, low_memory=False)
        except Exception as exc:  # noqa: BLE001
            last_error = exc
    raise RuntimeError(f"无法读取文件: {path} ({last_error})") from last_error


def read_json_file(path: Path) -> Any:
    with open(path, "r", encoding="utf-8") as file:
        return json.load(file)


@dataclass
class KnowledgeBaseEntry:
    specialty_name: str
    folder_name: str
    disease_catalog: Path
    drug_catalog: Path
    risk_rules: Path


class KnowledgeBaseIndex:
    def __init__(self, paths: ExperimentPaths) -> None:
        self.paths = paths
        self.index_df = read_csv_flexible(paths.kb_index_file)

    def get_entry(self, specialty_name: str) -> KnowledgeBaseEntry:
        row = self.index_df[self.index_df["specialty_name"] == specialty_name]
        if row.empty:
            raise KeyError(f"未找到专科知识库索引: {specialty_name}")
        item = row.iloc[0]
        folder_name = str(item["folder_name"])
        return KnowledgeBaseEntry(
            specialty_name=specialty_name,
            folder_name=folder_name,
            disease_catalog=self._resolve_index_path(item["disease_catalog"], folder_name),
            drug_catalog=self._resolve_index_path(item["drug_catalog"], folder_name),
            risk_rules=self._resolve_index_path(item["risk_rules"], folder_name),
        )

    def _resolve_index_path(self, raw_path: str, folder_name: str) -> Path:
        candidate = Path(str(raw_path))
        if candidate.exists():
            return candidate

        if not candidate.is_absolute():
            repo_candidate = self.paths.root_dir / candidate
            if repo_candidate.exists():
                return repo_candidate

        filename = PureWindowsPath(str(raw_path)).name
        return self.paths.knowledge_base_dir / folder_name / filename


class SpecialtyKnowledgeLoader:
    def __init__(self, kb_index: KnowledgeBaseIndex) -> None:
        self.kb_index = kb_index

    def load(self, specialty_name: str) -> dict[str, Any]:
        entry = self.kb_index.get_entry(specialty_name)
        disease_catalog = read_csv_flexible(entry.disease_catalog)
        drug_catalog = read_csv_flexible(entry.drug_catalog)
        risk_rules = read_json_file(entry.risk_rules)

        return {
            "entry": entry,
            "disease_catalog": disease_catalog,
            "drug_catalog": drug_catalog,
            "risk_rules": risk_rules,
        }

    def build_prompt_payload(self, specialty_name: str) -> dict[str, Any]:
        kb = self.load(specialty_name)
        disease_catalog = kb["disease_catalog"]
        drug_catalog = kb["drug_catalog"]

        disease_columns = [
            "disease_name",
            "aliases",
            "diagnostic_basis",
            "key_symptoms",
            "key_labs_or_tests",
            "differential_diagnosis",
            "reference_source",
            "reference_url",
            "agent_use",
        ]
        drug_columns = [
            "standard_drug_name",
            "aliases",
            "drug_class",
            "disease_context",
            "treatment_role",
            "order_category",
            "mechanism_or_function",
            "major_cautions",
            "reference_source",
            "reference_url",
            "agent_use",
        ]
        disease_payload = disease_catalog[
            [column for column in disease_columns if column in disease_catalog.columns]
        ].head(30)
        disease_directed_drugs = drug_catalog[
            drug_catalog["treatment_role"].isin(["disease_directed_therapy", "risk_modifying_therapy"])
        ][[column for column in drug_columns if column in drug_catalog.columns]].head(30)
        supportive_drugs = drug_catalog[
            drug_catalog["treatment_role"].isin(
                ["supportive_or_symptomatic_therapy", "general_inpatient_medication"]
            )
        ][[column for column in drug_columns if column in drug_catalog.columns]].head(20)

        return {
            "specialty_name": specialty_name,
            "knowledge_base_dir": str(kb["entry"].disease_catalog.parent),
            "diagnostic_knowledge": disease_payload.fillna("").to_dict(orient="records"),
            "drug_function_knowledge": {
                "disease_directed_or_risk_modifying": disease_directed_drugs.fillna("").to_dict(
                    orient="records"
                ),
                "supportive_or_general": supportive_drugs.fillna("").to_dict(orient="records"),
            },
            "risk_rules": kb["risk_rules"],
        }
